## Problem 23: Non-abundant Sums

A perfect number is a number for which the sum of its proper divisors is exactly equal to the number. For example, the sum of the proper divisors of $28$ would be $1 + 2 + 4 + 7 + 14 = 28$, which means that $28$ is a perfect number.

A number $n$ is called *deficient* if the sum of its proper divisors is less than $n$ and it is called *abundant* if this sum exceeds $n$.

As $12$ is the smallest abundant number, $1 + 2 + 3 + 4 + 6 = 16$, the smallest number that can be written as the sum of two abundant numbers is $24$. By mathematical analysis, it can be shown that all integers greater than $28123$ can be written as the sum of two abundant numbers. However, this upper limit cannot be reduced any further by analysis even though it is known that the greatest number that cannot be expressed as the sum of two abundant numbers is less than this limit.

Find the sum of all the positive integers which cannot be written as the sum of two abundant numbers.

In [1]:
''' 
Returns all the proper divisors for a number.

eg:
>>> proper_divisors(24)
>>> [1, 2, 3, 4, 6, 8, 12]
'''

def proper_divisors(number : int) -> list:
    '''
    --- Function Description --------------------------------------------------------------------------------------------------
        Returns all the proper divisors for a number.
    ---------------------------------------------------------------------------------------------------------------------------
    
    --- Function Inputs -------------------------------------------------------------------------------------------------------
        : list : number : The number to find the proper divisors for.
    ---------------------------------------------------------------------------------------------------------------------------
    
    --- Function Outputs ------------------------------------------------------------------------------------------------------
        : list : divisors : The proper divisors of the number.
    ---------------------------------------------------------------------------------------------------------------------------
    
    --- Function Examples -----------------------------------------------------------------------------------------------------
        >>> proper_divisors(24)
        >>> [1, 2, 3, 4, 6, 8, 12]
    ---------------------------------------------------------------------------------------------------------------------------
    '''
    
    # Check types of function inputs:
    if not isinstance(number, int): raise ValueError('Please enter an integer >1 for the number argument.')
    if number < 2: raise ValueError('Please enter an integer >1 for the number argument.')
    
    divisors = []
    
    for test_divisor in range(1, (number//2) + 1):
        
        if number % test_divisor == 0: divisors.append(test_divisor)
    
    return divisors

In [2]:
''' 
Tests if a number is abundant.

eg:
>>> is_abundant(12)
>>> True
'''

def is_abundant(number : int) -> bool:
    '''
    --- Function Description --------------------------------------------------------------------------------------------------
        Tests if a number is abundant.
    ---------------------------------------------------------------------------------------------------------------------------
    
    --- Function Inputs -------------------------------------------------------------------------------------------------------
        : int : number : The number to test for abundancy.
    ---------------------------------------------------------------------------------------------------------------------------
    
    --- Function Outputs ------------------------------------------------------------------------------------------------------
        : bool : The number is abundant.
    ---------------------------------------------------------------------------------------------------------------------------
    
    --- Function Examples -----------------------------------------------------------------------------------------------------
        >>> is_abundant(12)
        >>> True
    ---------------------------------------------------------------------------------------------------------------------------
    '''
    
    # Check types of function inputs:
    if not isinstance(number, int): raise ValueError('Please enter an integer >1 for the number argument.')
    if number < 2: raise ValueError('Please enter an integer >1 for the number argument.')
    
    divisor_sum = sum(proper_divisors(number))
    
    return divisor_sum > number

In [3]:
''' 
Returns a list of abundant numbers up to max_number.

eg:
>>> create_abundant_list(100)
>>> [12, 18, 20, 24, 30, 36, 40, 42, 48, 54, 56, 60, 66, 70, 72, 78, 80, 84, 88, 90, 96]
'''

def create_abundant_list(max_number : int) -> list:
    '''
    --- Function Description --------------------------------------------------------------------------------------------------
        Returns a list of abundant numbers up to max_number.
    ---------------------------------------------------------------------------------------------------------------------------
    
    --- Function Inputs -------------------------------------------------------------------------------------------------------
        : list : max_number : The maximum number to test.
    ---------------------------------------------------------------------------------------------------------------------------
    
    --- Function Outputs ------------------------------------------------------------------------------------------------------
        : list : abundant_numbers : The list of abundant numbers.
    ---------------------------------------------------------------------------------------------------------------------------
    
    --- Function Examples -----------------------------------------------------------------------------------------------------
        >>> create_abundant_list(100)
        >>> [12, 18, 20, 24, 30, 36, 40, 42, 48, 54, 56, 60, 66, 70, 72, 78, 80, 84, 88, 90, 96]
    ---------------------------------------------------------------------------------------------------------------------------
    '''
    
    # Check types of function inputs:
    if not isinstance(max_number, int): raise ValueError('Please enter an integer >1 for the max_number argument.')
    if max_number < 2: raise ValueError('Please enter an integer >1 for the max_number argument.')
    
    abundant_numbers = []
  
    for test_number in range(12, max_number):
        
        if is_abundant(test_number): 
            abundant_numbers.append(test_number)
    
    return abundant_numbers

In [4]:
''' 
Returns a list of numbers that are not expressible as a sum of two abundant numbers up to max_number.

eg:
>>> create_not_expressible_list(25)
>>> [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 25]
'''

def create_not_expressible_list(max_number : int) -> list:
    '''
    --- Function Description --------------------------------------------------------------------------------------------------
        Returns a list of numbers that are not expressible as a sum of two abundant numbers up to max_number.
    ---------------------------------------------------------------------------------------------------------------------------
    
    --- Function Inputs -------------------------------------------------------------------------------------------------------
        : list : max_number : The list of names to sort.
    ---------------------------------------------------------------------------------------------------------------------------
    
    --- Function Outputs ------------------------------------------------------------------------------------------------------
        : list : not_expressible_numbers : The list of numbers that are not expressible as a sum of two abundant numbers.
    ---------------------------------------------------------------------------------------------------------------------------
    
    --- Function Examples -----------------------------------------------------------------------------------------------------
        >>> create_not_expressible_list(25)
        >>> [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 25]
    ---------------------------------------------------------------------------------------------------------------------------
    '''
    
    # Check types of function inputs:
    if not isinstance(max_number, int): raise ValueError('Please enter an integer >1 for the max_number argument.')
    if max_number < 2: raise ValueError('Please enter an integer >1 for the max_number argument.')
    
    all_abundant_numbers = create_abundant_list(max_number)
    
    expressible_list = []
    
    for number in range(max_number):
        expressible_list.append(False)
    
    for number_i in all_abundant_numbers:
        for number_j in all_abundant_numbers:
            if number_i + number_j - 1 < len(expressible_list):
                expressible_list[number_i + number_j - 1] = True
    
    not_expressible_numbers = []
    
    for index, expressible_bool in enumerate(expressible_list):
        if not expressible_bool: not_expressible_numbers.append(index+1)
    
    return not_expressible_numbers

In [5]:
problem_max_number = 28123

solution = create_not_expressible_list(problem_max_number)

In [6]:
print(f'The largest number not expressible by the sum of two abundant numbers is {solution[len(solution)-1]:,}')
print(f'and the sum of all numbers not expressible by the sum of two abundant numbers is {sum(solution):,}.')

The largest number not expressible by the sum of two abundant numbers is 20,161
and the sum of all numbers not expressible by the sum of two abundant numbers is 4,179,871.


### Problem 23 Solution:

The largest number not expressible by the sum of two abundant numbers is 20,161 and the sum of all numbers not expressible by the sum of two abundant numbers is 4,179,871.